extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -36,30 +44,54 @@ class AmpServiceWorkerInstallation extends AMP.BaseElement {
     const src = this.element.getAttribute('src');
     assertHttpsUrl(src, this.element);
 
+    /** @private {?string}  */
+    this.iframeSrc_ = null;
+
     if (isProxyOrigin(src) || isProxyOrigin(win.location.href)) {
+      const iframeSrc = this.element.getAttribute('data-iframe-src');
+      if (iframeSrc) {
+        assertHttpsUrl(iframeSrc, this.element);
+        const origin = parseUrl(iframeSrc).origin;
+        const sourceOrigin = getSourceOrigin(win.location.href);
+        user.assert(origin == sourceOrigin,
+            'data-iframe-src should be a URL on the same origin as the ' +
+            'source origin of the AMP document. %s VS. %s',
+            origin, sourceOrigin);
+        this.iframeSrc_ = iframeSrc;
+      }
       return;
     }
 
-    if (originMatches(win.location.href, src)) {
+    if (parseUrl(win.location.href).origin == parseUrl(src).origin) {
       install(this.getWin(), src);
     } else {
-      if (getMode().development) {
-        user.warn(TAG,
-            'Did not install ServiceWorker because it does not ' +
-            'match the current origin: ' + src);
-      }
+      user.error(TAG,
+          'Did not install ServiceWorker because it does not ' +
+          'match the current origin: ' + src);
     }
   }
-}
 
-/**
- * Returns true if the 2 hrefs are on the same origin.
- * @param {string} href1
- * @param {string} href2
- * return {boolean}
- */
-function originMatches(href1, href2) {
-  return parseUrl(href1).origin == parseUrl(href2).origin;
+  /** @override */
+  renderOutsideViewport() {
+    return false;
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (!this.iframeSrc_) {
+      return Promise.resolve();
+    }
+    this.element.style.display = 'none';
```

Comment:
The element is already with `layout=nodisplay`, right? Then how does its `layoutCallback` ever called in the first place?